File: solutions-by-language/java/1504_count_submatrices_with_all_ones.java
import java.util.*;

class Solution {
    public int numSubmat(int[][] mat) {
        int m = mat.length, n = mat[0].length;
        int[] heights = new int[n];
        int total = 0;

        for (int i = 0; i < m; i++) {
            // update histogram heights
            for (int j = 0; j < n; j++) {
                if (mat[i][j] == 1)
                    heights[j] += 1;
                else
                    heights[j] = 0;
            }

            // monotonic stack
            int[] rowSum = new int[n];
            Deque<Integer> stack = new ArrayDeque<>();
            for (int j = 0; j < n; j++) {
                while (!stack.isEmpty() && heights[stack.peek()] >= heights[j]) {
                    stack.pop();
                }

                if (!stack.isEmpty()) {
                    int prev = stack.peek();
                    rowSum[j] = rowSum[prev] + heights[j] * (j - prev);
                } else {
                    rowSum[j] = heights[j] * (j + 1);
                }

                stack.push(j);
                total += rowSum[j];
            }
        }
        return total;
    }
}

File: solutions-by-language/python/1504_count_submatrices_with_all_ones.py
class Solution:
    def numSubmat(self, mat: List[List[int]]) -> int:
        m, n = len(mat), len(mat[0])
        heights = [0] * n
        total = 0

        for i in range(m):
            for j in range(n):
                if mat[i][j] == 1:
                    heights[j] += 1
                else:
                    heights[j] = 0

            stack = []
            row_sum = [0] * n
            for j in range(n):
                while stack and heights[stack[-1]] >= heights[j]:
                    stack.pop()

                if stack:
                    prev = stack[-1]
                    row_sum[j] = row_sum[prev] + heights[j] * (j - prev)
                else:
                    row_sum[j] = heights[j] * (j + 1)

                stack.append(j)
                total += row_sum[j]

        return total